V7 Caching issue

mobileWMS opened the pull request:
Main goal of these changes was to fix a caching issue, where browsers kept giving the same exact data in the exported Excel or CSV even if we did a new search and loaded new data.

Changes:

1. In exporter-addon, made changes to fix caching and improve file name
2. In exporter-demo, made changes so tester could add lines to table, see changes in exported file
3. In both, upgrade to latest public Vaadin 7 and other minor version updates, and changed project version.

Changes might be helpful for Vaadin 8, not needed for Flow ( as far as I know ).

mobileWMS committed:
added fixes to resolve cache issues when download again

When downloading from the same table, after making changes to table (
loading new data or adding new lines ), it would keep using the same
exact file because of caching.  So set the stream cache to 0 to turn off
this caching.

Also added some better file naming conventions.  If CSVExporter, and
download file name ends in ".csv", use as is ( no change ).  Else, add
date/time stamp to make file name unique.  Same with ExcelExporter, but
use file name as is if it ends in ".xls" ( no change ), adding  time
stamp otherwise.  In both cases, in `Exporter.setContainerToBeExported(
container );`, we call `setDownloadFileName(getDownloadFileName);` to
refresh the file name automatically when data is loaded into the
exporter.

I have found that in Chrome, if I set the file name for "Excel", such as
"demo-excel.xls", it would keep using the cache.  As such, best case is
to use "demo-excel" ( without a file suffix ), seems to be more
realiable.

mobileWMS committed:
Added more flexibility to demo

Added ability to add lines to table so that we could see exported files
change

haiwan commented on exporter-addon/pom.xml:
```diff
@@ -5,14 +5,14 @@
-		<maven.compiler.target>1.6</maven.compiler.target>
-		<vaadin.version>7.6.6</vaadin.version>
+		<maven.compiler.source>1.8</maven.compiler.source>
+		<maven.compiler.target>1.8</maven.compiler.target>
```
Officially, Vaadin 7 supports Java 1.6, so maybe we keep it as 1.6?

mobileWMS replied:
Good point, will make change.  Strictly speaking, we don't need vaadin version 7.7.13 for my change.  Personally, I like keeping it, but do you think we should keep it at vaadin version 7.6.6, or does not not really matter?  Since I need to update Java version, thought I would ask.

haiwan replied:
Vaadin version doesn't matter here, Vaadin dependency actually should be in **provided** scope.

haiwan commented on exporter-addon/src/main/java/org/vaadin/haijian/CSVExporter.java:
```diff
@@ -28,15 +34,26 @@ protected FileBuilder createFileBuilder(Container container) {
+    		fileDateTimeFmt = new SimpleDateFormat("yyyyMMddHHmmss");
+    	return fileDateTimeFmt;
+    }
+    
```
Some duplicated code here in CSVExporter and ExcelExporter, maybe you can move the fileDateTimeFmt and the getter method to the parent Exporter class?

Also in the getDateTimeFmt() method, it could be `if( fileDateTimeFmt == null))` ? instead of `if( !( fileDateTimeFmt instanceof SimpleDateFormat ))`?

mobileWMS replied:
Good point, will do.

haiwan commented on exporter-demo/pom.xml:
```diff
@@ -5,14 +5,14 @@
-		<maven.compiler.target>1.6</maven.compiler.target>
-		<vaadin.version>7.6.6</vaadin.version>
+		<maven.compiler.source>1.8</maven.compiler.source>
+		<maven.compiler.target>1.8</maven.compiler.target>
```
same here, let's keep it as 1.6?

mobileWMS commented:
Stress tested it in my application and seems to work well.  So it works in demo, and works in quite a few cases for my production app ( if not all cases ).

mobileWMS committed:
applied changes suggested by haijian

changed java version back to 1.6
moved some common code to parent class.

mobileWMS committed:
changes suggested by haijian

changed java version from 1.8 back to 1.6
added 2 new buttons so could see file with date/time and without.

haiwan commented:
Good job!
